ArdusatSDK.cpp

```diff
@@ -453,266 +217,1101 @@ const char * valuesToCSV(const char *sensorName, unsigned long timestamp, int nu
+ * @ingroup magnetic
+ */
+void Magnetic::read(void) {
+  this->header.timestamp = millis();
+  lsm303_getMag(&(this->x), &(this->y), &(this->z));
+}
+
+/**
+ * @brief   Returns last read value in CSV format
+ * @ingroup magnetic
+ * @param   sensorName The text to display next to the value
+ * @return  sensor readings in CSV format or empty string if uninitialized
+ */
+const char * Magnetic::toCSV(const char * sensorName) {
+  if (this->initialized) {
+    return valuesToCSV(sensorName, this->header.timestamp, 3,
+                       this->x, this->y, this->z);
+  } else {
+    return "";
+  }
+}
+
+/**
+ * @brief   Returns last read value in JSON format
+ * @ingroup magnetic
+ * @param   sensorName The text to display next to the value
+ * @return  sensor readings in JSON format or empty string if uninitialized
+ */
+const char * Magnetic::toJSON(const char * sensorName) {
+  if (this->initialized) {
+    return valuesToJSON(sensorName, this->header.unit, 3, "X", this->x,
+                              "Y", this->y, "Z", this->z);
+  } else {
+    return "";
+  }
+}
+
+
+/**************************************************************************//**
+ * @brief   Constructs Orientation calculation object
+ * @ingroup orientation
+ *
+ * Example Usage:
+ * @code
+ *     Orientation orient;                               // Instantiate sensor object
+ *     orient.begin();                                   // Initialize sensor
+ *     Serial.println(orient.readToJSON("orientation")); // Read and print values in JSON
+ * @endcode
+ *****************************************************************************/
+Orientation::Orientation(void) {
+  Acceleration accel;
+  Magnetic mag;
+  this->accel = &accel;
+  this->mag = &mag;
+
+  this->initializeHeader(SENSORID_ADAFRUIT9DOFIMU, DATA_UNIT_DEGREES, orientation_sensor_name);
+}
+
+/**
+ * @brief   Uses provided Acceleration and Magnetic Sensors to construct Orientation calculation object
+ * @ingroup orientation
+ *
+ * Example Usage:
+ * @code
+ *     Acceleration accel;
+ *     Magnetic mag;
+ *     Orientation orient(accel, mag);                   // Instantiate sensor object
+ *     orient.begin();                                   // Initialize sensor
+ *     Serial.println(orient.readToJSON("orientation")); // Read and print values in JSON
+ * @endcode
+ */
+Orientation::Orientation(Acceleration & accel, Magnetic & mag) :
+  accel(&accel),
+  mag(&mag)
 {
-  _resetOutBuf();
-  _writeJSONValue(_getOutBuf(), sensor_name, unit_to_str(lum.header.unit), lum.lux);
-  return _getOutBuf();
+  this->initializeHeader(SENSORID_ADAFRUIT9DOFIMU, DATA_UNIT_DEGREES, orientation_sensor_name);
+}
+
+/**
+ * @brief   Initializes the sensor with any set configurations or defaults
+ * @ingroup orientation
+ *
+ * @retval true  Successfully initialized
+ * @retval false Failed to initialize
+ */
+boolean Orientation::initialize(void) {
+  return accel->begin() && mag->begin();
```

Comment:
I'm fairly certain it's not a big deal, but have you tested to make sure this works (e.g. gives you good values) evne when both `accel` and `mag` are already initialized? (e.g. that calling `begin()` twice on these sensors doesn't cause issues)